modules/transformer-batch/src/main/scala/com/snowplowanalytics/snowplow/rdbloader/transformer/batch/badrows/PartitionDataFilter.scala

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright (c) 2012-2023 Snowplow Analytics Ltd. All rights reserved.
+ *
+ * This program is licensed to you under the Apache License Version 2.0,
+ * and you may not use this file except in compliance with the Apache License Version 2.0.
+ * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+package com.snowplowanalytics.snowplow.rdbloader.transformer.batch.badrows
+
+import com.snowplowanalytics.snowplow.rdbloader.common.transformation.Transformed
+
+object PartitionDataFilter {
+
+  type PartitionData = Iterator[Transformed]
+  type GoodData = Iterator[Transformed]
+  type BadData = Iterator[Transformed]
+
+  def extractGoodAndSinkBad(
+    data: PartitionData,
+    partitionIndex: Int,
+    badrowSink: BadrowSink
+  ): GoodData = {
+    val (good, bad) = splitPartitionToGoodAndBad(data)
+
+    if (bad.hasNext) {
+      val badrowAsStrings = bad
+        .collect { case Transformed.WideRow(false, dString) => dString.value }
+
+      badrowSink.sink(badrowAsStrings, partitionIndex)
```

Comment:
> bad rows are sunk directly from the different workers (EMR nodes) where the RDD[Transformed]'s partitions are processed

Yes, thats exactly what's happening. I introduced this approach with mapping partitions initially only for parquet badrows [here](https://github.com/snowplow/snowplow-rdb-loader/commit/c27557700887f2b60e382c3ff39ede4056d99e32) so we don't have to use `.cache()` on Spark RDD.

 I thought it would be also the best way to add Kinesis sink. I can't think of any different way to add Kinesis and make it work with Spark and all this RDD stuff :/